pkg/component/registrycaches/registrycaches.go

```diff
@@ -0,0 +1,257 @@
+	namespace string,
+	values Values,
+) component.DeployWaiter {
+	return &registryCaches{
+		client:    client,
+		namespace: namespace,
+		values:    values,
+	}
+}
+
+type registryCaches struct {
+	client    client.Client
+	namespace string
+	values    Values
+}
+
+// Deploy implements component.DeployWaiter.
+func (r *registryCaches) Deploy(ctx context.Context) error {
+	data, err := r.computeResourcesData()
+	if err != nil {
+		return err
+	}
+
+	var (
+		origin      = "registry-cache"
+		keepObjects = false
+
+		secretName, secret = managedresources.NewSecret(r.client, r.namespace, ManagedResourceName, data, false)
+		managedResource    = managedresources.NewForShoot(r.client, r.namespace, ManagedResourceName, origin, keepObjects).
+					WithSecretRef(secretName).
+					DeletePersistentVolumeClaims(true)
+	)
+
+	if err := secret.Reconcile(ctx); err != nil {
+		return fmt.Errorf("failed to create or update secret of managed resources: %w", err)
+	}
+
+	if err := managedResource.Reconcile(ctx); err != nil {
+		return fmt.Errorf("failed to not create or update managed resource: %w", err)
+	}
+
+	return nil
+}
+
+// Destroy implements component.DeployWaiter.
+func (r *registryCaches) Destroy(ctx context.Context) error {
+	return managedresources.Delete(ctx, r.client, r.namespace, ManagedResourceName, false)
+}
+
+// TimeoutWaitForManagedResource is the timeout used while waiting for the ManagedResources to become healthy
+// or deleted.
+var TimeoutWaitForManagedResource = 2 * time.Minute
+
+// Wait implements component.DeployWaiter.
+func (r *registryCaches) Wait(ctx context.Context) error {
+	timeoutCtx, cancel := context.WithTimeout(ctx, TimeoutWaitForManagedResource)
+	defer cancel()
+
+	return managedresources.WaitUntilHealthy(timeoutCtx, r.client, r.namespace, ManagedResourceName)
+}
+
+// WaitCleanup implements component.DeployWaiter.
+func (r *registryCaches) WaitCleanup(ctx context.Context) error {
+	timeoutCtx, cancel := context.WithTimeout(ctx, TimeoutWaitForManagedResource)
+	defer cancel()
+
+	return managedresources.WaitUntilDeleted(timeoutCtx, r.client, r.namespace, ManagedResourceName)
+}
+
+func (r *registryCaches) computeResourcesData() (map[string][]byte, error) {
+	objects := []client.Object{
+		&corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: constants.NamespaceRegistryCache,
+			},
+		},
+	}
+
+	for _, cache := range r.values.Caches {
+		cacheObjects, err := computeResourcesDataForRegistryCache(&cache, r.values.Image)
+		if err != nil {
+			return nil, fmt.Errorf("failed to compute resources for upstream %s: %w", cache.Upstream, err)
+		}
+
+		objects = append(objects, cacheObjects...)
+	}
+
+	registry := managedresources.NewRegistry(kubernetes.ShootScheme, kubernetes.ShootCodec, kubernetes.ShootSerializer)
+
+	return registry.AddAllAndSerialize(objects...)
+}
+
+func computeResourcesDataForRegistryCache(cache *v1alpha1.RegistryCache, image string) ([]client.Object, error) {
+	if cache.Size == nil {
```

Comment:
I tried to follow defensive programming paradigm to fail with a more meaningful error instead of a panic (would cause a crash of the whole process) if the defaulting or validation (in registry-cache-admission) don't work for some reason.

Reply:
Alright, I just remember from some reviews that we did in g/g that we avoid such checks when there is validation/defaulting, but I don't really mind tbh.